BB-295 - bootstrap ts compilation

jbertran opened the pull request:
Outcome so far:

* TypeScript is integrated in the build environment, and compiles to the dist folder
* Jest is configured to use the mixed sources with on-the-fly compilation, and runs all tests that were run with mocha
* Some test scripts pass:
  * unit
  * routes
  * retry

Remaining work is simply ensuring all other test suites pass (bonus: make them separate parallel steps in the build process, such that the work doesn’t need to be sequential).

Zenko Build: https://github.com/scality/Zenko/actions/runs/3822509132

Commit:
BB-295 - introduce typescript and jest

Commit:
BB-295 - use compiled output in tests

Commit:
BB-295 - catch up jest CLI differences

Commit:
BB-295 - fix eslint

- add eslint-plugin-jest for jest globals
- upgrade esling
- fix remaining lint issues

Commit:
BB-295 - use ts-jest

Commit:
BB-295 - streamline syntheticbucketd container build

- provide minimal package.json
- move all files to tests/utils/syntheticbucketd

Commit:
BB-295 - no parallelism in tests

Commit:
BB-295 - adapt to jest API

Commit:
BB-295 - switch to jest

Commit:
BB-295 - stop leaking redis connections in BBAPI tests

Commit:
BB-295 - stop leaking redis connection in qp tests

Commit:
BB-295 - sample migration

Commit:
BB-295 - fix eslint with ts files

Commit:
BB-295 - move tests to .spec.js

Commit:
BB-295 - fix missed tests

Commit:
BB-295 - use jest with expression filters instead of paths

Commit:
BB-295 - set forceExit on every jest launch

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-295 contains:


* `8.3.20`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.5.0`


Please check the `Fix Version/s` of BB-295, or the target
branch of this pull request.

jbertran commented:
/approve

jbertran commented:
/wait

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-295 contains:


* `8.3.20`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.5.0`


Please check the `Fix Version/s` of BB-295, or the target
branch of this pull request.




*The following options are set:* **approve**

Commit:
BB-295 - fix image build issue

Commit:
BB-295 - fix functional tests

francoisferrand commented on Dockerfile:
```diff
@@ -28,11 +28,12 @@ RUN wget https://github.com/jwilder/dockerize/releases/download/$DOCKERIZE_VERSI
     && rm dockerize-linux-amd64-$DOCKERIZE_VERSION.tar.gz
 
-COPY package.json yarn.lock /usr/src/app/
+COPY . /usr/src/app
```
the goal here is to perform yarn install first, without depending on changing to the code to allow better layer caching : hence it is required to copy only package.json and yarn.lock

To deal with the build, you should thus do it in 2 steps here:

```
COPY package.json yarn.lock /usr/src/app/
RUN yarn install ...
COPY . /usr/src/app
RUN yarn build
```

francoisferrand commented on Dockerfile:
```diff
@@ -45,7 +46,8 @@ RUN apt-get update && \
 
 # Keep the .git directory in order to properly report version
-COPY . /usr/src/app
+COPY . ./
```
not needed? since data is copied from `dist/` as well..

tmacro replied:
Only thing that needs to be copied is package.json if we intend to use yarn to launch things
```suggestion
COPY ./package.json ./
```

tmacro commented on tsconfig.json:
```diff
@@ -0,0 +1,32 @@
+{
```
One of the things I found while standing up Scuba's TS config was https://www.npmjs.com/package/@tsconfig/node16
It basically just a base you can extend, tuned for a particular use case.
I used it here https://github.com/scality/scuba/blob/development/1.0/tsconfig.json
The options can be set manually of course, but I think it's nice to have a maintained base to build off of.

tmacro commented on tsconfig.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "compilerOptions": {
+      "target": "es6",
```
I'd suggest setting this to `es2021`. Among other things it uses native node promises instead of a polyfill.
```suggestion
      "target": "es2021",
```

tmacro commented on tsconfig.json:
```diff
@@ -0,0 +1,32 @@
+      "noImplicitAny": false,
+      "noEmitOnError": false,
+      "sourceMap": true,
+      "skipLibCheck": true
```
Need to add the es2021 lib to use the target
```suggestion
      "skipLibCheck": true,
      "lib": ["es2021"]
```

tmacro commented on package.json:
```diff
@@ -61,21 +64,40 @@
     "node-rdkafka": "^2.12.0",
     "node-schedule": "^1.2.0",
     "node-zookeeper-client": "^1.1.3",
+    "typescript": "^4.8.4",
```
This should be listed as a dev dependency

tmacro commented on package.json:
```diff
@@ -44,6 +46,7 @@
   },
   "homepage": "https://github.com/scality/backbeat#readme",
   "dependencies": {
+    "@types/node": "^18.11.9",
```
This should be listed as a dev dependency

tmacro commented on Dockerfile:
```diff
@@ -28,11 +28,12 @@ RUN wget https://github.com/jwilder/dockerize/releases/download/$DOCKERIZE_VERSI
 
-COPY package.json yarn.lock /usr/src/app/
+COPY . /usr/src/app
 RUN yarn install --ignore-engines --frozen-lockfile --production --network-concurrency 1 \
     && rm -rf /var/lib/apt/lists/* \
     && rm -rf ~/.node-gyp \
     && rm -rf /tmp/yarn-*
+RUN yarn build
```
When typescript is moved to a dev dependency this needs to be split if you want to continue copying the node_modules dir from the builder.
Also we can get rid of the `rm`s since the builder image will be trashed at the end anyway.

```suggestion
RUN yarn install --ignore-engines --frozen-lockfile --network-concurrency 1

RUN yarn build

RUN rm -rf node_modules \
    && yarn install --production --frozen-lockfile --ignore-optional
```

rachedbenmustapha commented on tests/functional/replication/queueProcessor.spec.js:
```diff
@@ -877,9 +877,9 @@ describe('queue processor functional tests with mocking', () => {
+    afterAll(done => {
         httpServer.close();
-        async.parallel([
+        async.series([
```
This doesn't seem to be strictly related to typescript/jest, is there a reason for this change?

rachedbenmustapha commented on tests/functional/notification/NotificationConfigManager.spec.js:
```diff
@@ -42,7 +57,7 @@ describe('NotificationConfigManager ::', () => {
         const getCollectionStub = sinon.stub().returns({
             // mock change stream
-            watch: () => new events.EventEmitter(),
+            watch: () => new MockChangeStream(),
```
This doesn't seem to be strictly related to typescript/jest, is there a reason for this change?

rachedbenmustapha commented on tests/unit/OffsetLedger.spec.js:
```diff
@@ -1,6 +1,6 @@
 const assert = require('assert');
 
-const OffsetLedger = require('../../lib/OffsetLedger');
+const OffsetLedger = require('../../lib/OffsetLedger').default;
```
Naming `default` explicitly feels off; isn't there some more typescript-native way of importing the default export without explicitly naming it?

francoisferrand replied:
I think this is the way, the same is done in arsenal.

Maybe this is due to the dynamic transpiling of library's TS code, in the tests...
And the proper solution would be to migrate the test to typescript (or possibly use ES6 `import` instead of `require`), but this is not in scope of this ticket: which only deals with "bootstraping", i.e. having the toolchain in place so that it is _possible_ to start writing components in TS)

rachedbenmustapha commented on tests/utils/syntheticbucketd/Dockerfile:
```diff
@@ -1,10 +1,7 @@
 RUN mkdir /app
-COPY package.json /app
-COPY yarn.lock /app
+COPY * /app
```
Why this change? It invalidates the layer if anything changes even if unrelated to dependencies, and triggers an avoidable `yarn install`.

rachedbenmustapha commented on tests/functional/lib/BackbeatConsumer.spec.js:
```diff
@@ -122,7 +121,7 @@ describe('BackbeatConsumer main tests', () => {
                     _checkZkMetrics(() => {
                         consumeCb();
-                        consumer.unsubscribe();
+                        consumer.pause();
```
Why this change (and the `subscribe`/`resume` one)? Was the test wrong before?

rachedbenmustapha commented on tests/unit/RoleCredentials.spec.js:
```diff
@@ -55,100 +55,133 @@ function _assertCredentials(err, roleCredentials, cb) {
         roleCredentials.get(err => {
             if (err) {
-                return done(err);
+                done(err);
```
I think the missing return changes the test's flow.

rachedbenmustapha commented on tests/unit/RoleCredentials.spec.js:
```diff
@@ -55,100 +55,133 @@ function _assertCredentials(err, roleCredentials, cb) {
+    it('should refresh credentials upon expiration', done => {
         roleCredentials.get(err => {
             if (err) {
-                return done(err);
```
I think the missing return changes the test's flow.

rachedbenmustapha commented on tests/unit/RoleCredentials.spec.js:
```diff
@@ -55,100 +55,133 @@ function _assertCredentials(err, roleCredentials, cb) {
+    it('should use same credentials if not expired or about to expire', done => {
         roleCredentials.get(err => {
             if (err) {
-                return done(err);
```
I think the missing return changes the test's flow.

rachedbenmustapha commented on lib/api/BackbeatAPI.js:
```diff
@@ -101,31 +101,31 @@ class BackbeatAPI {
             internalStart: this._internalStart,
         };
-        const metrics = new Metrics(metricsConfig, logger);
+        this._metrics = new Metrics(metricsConfig, logger);
```
Why is this needed?

francoisferrand replied:
the objects needs to be 'stored' so the resources can be closed (in order not to leak) at the end of the tests:

```
bbapi._metrics.disconnect();
```

francoisferrand committed:
Fix syntheticbucketd dockerfile to support caching deps in layer

Issue: BB-295

Seragonia committed:
BB-400: bump mongodb to 4.4.21

bert-e committed:
Merge branch 'improvement/BB-424-bump-8.6.24' into tmp/octopus/w/8.7/improvement/BB-424-bump-8.6.24

bert-e committed:
Merge branch 'bugfix/BB-415' into tmp/octopus/w/8.7/bugfix/BB-415

bert-e committed:
Merge branch 'bugfix/BB-366/addGCReadAccountIDs' into tmp/octopus/w/8.7/bugfix/BB-366/addGCReadAccountIDs

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-415' and 'q/2424/8.6/bugfix/BB-415' into tmp/octopus/q/8.7

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-366/addGCReadAccountIDs' and 'q/2410/8.6/bugfix/BB-366/addGCReadAccountIDs' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-425/order' into tmp/octopus/w/8.7/bugfix/BB-425/order

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-425/bump' into tmp/octopus/w/8.7/bugfix/BB-425/bump

jonathan-gramain committed:
Merge remote-tracking branch 'origin/w/8.6/bugfix/BB-419-increaseNotificationDefaultConcurrency' into w/8.7/bugfix/BB-419-increaseNotificationDefaultConcurrency

bert-e committed:
Merge branch 'improvement/BB-427-remove-zero-byte-skip-for-dmf' into tmp/octopus/w/8.7/improvement/BB-427-remove-zero-byte-skip-for-dmf

bert-e committed:
Merge branch 'feature/BB-189-support-archived-object-delete' into tmp/octopus/w/8.7/feature/BB-189-support-archived-object-delete

bert-e committed:
Merge branches 'w/8.7/feature/BB-189-support-archived-object-delete' and 'q/2432/8.6/feature/BB-189-support-archived-object-delete' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'feature/BB-429' into tmp/octopus/w/8.7/feature/BB-429

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-431/get-backbeat-client' into tmp/octopus/w/8.7/bugfix/BB-431/get-backbeat-client

bert-e committed:
Merge branch 'w/8.6/improvement/BB-428/lifecycle-listings' into tmp/octopus/w/8.7/improvement/BB-428/lifecycle-listings

bert-e committed:
Merge branch 'improvement/BB-430-improve-oplog-rule-lifecycle' into tmp/octopus/w/8.7/improvement/BB-430-improve-oplog-rule-lifecycle

bert-e committed:
Merge branches 'w/8.7/improvement/BB-430-improve-oplog-rule-lifecycle' and 'q/2438/8.6/improvement/BB-430-improve-oplog-rule-lifecycle' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/improvement/BB-433/bump' into tmp/octopus/w/8.7/improvement/BB-433/bump

bert-e committed:
Merge branch 'improvement/BB-432-remove-useless-kafka-connectors' into tmp/octopus/w/8.7/improvement/BB-432-remove-useless-kafka-connectors

bert-e committed:
Merge branch 'w/8.6/improvement/BB-434/newerNoncurrentVersions' into tmp/octopus/w/8.7/improvement/BB-434/newerNoncurrentVersions

bert-e committed:
Merge branch 'w/8.6/improvement/BB-434/newerNoncurrentVersions' into tmp/octopus/w/8.7/improvement/BB-434/newerNoncurrentVersions

bert-e committed:
Merge branch 'bugfix/BB-435' into tmp/octopus/w/8.7/bugfix/BB-435

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-435' and 'q/2444/8.6/bugfix/BB-435' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/feature/BB-413-circuit-breaker-backport' into tmp/octopus/w/8.7/feature/BB-413-circuit-breaker-backport

bert-e committed:
Merge branch 'w/8.6/feature/BB-413-circuit-breaker-backport' into tmp/octopus/w/8.7/feature/BB-413-circuit-breaker-backport

bert-e committed:
Merge branch 'w/8.6/feature/BB-413-circuit-breaker-backport' into tmp/octopus/w/8.7/feature/BB-413-circuit-breaker-backport

bert-e committed:
Merge branch 'improvement/BB-436' into tmp/octopus/w/8.7/improvement/BB-436

bert-e committed:
Merge branches 'w/8.7/improvement/BB-436' and 'q/2445/8.6/improvement/BB-436' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'improvement/BB-438' into tmp/octopus/w/8.7/improvement/BB-438

bert-e committed:
Merge branch 'bugfix/BB-448/could-not-get-location-configuration-error-log' into tmp/octopus/w/8.7/bugfix/BB-448/could-not-get-location-configuration-error-log

bert-e committed:
Merge branch 'bugfix/BB-437' into tmp/octopus/w/8.7/bugfix/BB-437

bert-e committed:
Merge branch 'improvement/BB-439' into tmp/octopus/w/8.7/improvement/BB-439

bert-e committed:
Merge branch 'w/8.6/feature/BB-451-delete-notification-fields' into tmp/octopus/w/8.7/feature/BB-451-delete-notification-fields

bert-e committed:
Merge branches 'w/8.7/improvement/BB-439' and 'q/2451/8.6/improvement/BB-439' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'bugfix/BB-445' into tmp/octopus/w/8.7/bugfix/BB-445

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-445' and 'q/2452/8.6/bugfix/BB-445' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-450/replication' into tmp/octopus/w/8.7/bugfix/BB-450/replication

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-450/replication' into tmp/octopus/w/8.7/bugfix/BB-450/replication

bert-e committed:
Merge branch 'improvement/BB-443' into tmp/octopus/w/8.7/improvement/BB-443

bert-e committed:
Merge branch 'w/8.6/improvement/BB-450/bump' into tmp/octopus/w/8.7/improvement/BB-450/bump

bert-e committed:
Merge branch 'improvement/BB-453-call-getmetadata-after-error-consumer' into tmp/octopus/w/8.7/improvement/BB-453-call-getmetadata-after-error-consumer

bert-e committed:
Merge branches 'w/8.7/improvement/BB-453-call-getmetadata-after-error-consumer' and 'q/2455/8.6/improvement/BB-453-call-getmetadata-after-error-consumer' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'bugfix/BB-459-forward-signals' into tmp/octopus/w/8.7/bugfix/BB-459-forward-signals

Commit:
Remove default behaviour of selecting the first cold location on expiration of cold objects

We don’t need this logic anymore as sorbet always sets the location name
and this might introduce unwanted behaviours when having multiple locations

Issue: BB-417

Commit:
add lint rule to avoid exclusive tests

Issue: BB-417

bert-e committed:
Merge branch 'bugfix/BB-457-oplog-disabled-cold-transition' into tmp/octopus/w/8.7/bugfix/BB-457-oplog-disabled-cold-transition

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-457-oplog-disabled-cold-transition' and 'q/2460/8.6/bugfix/BB-457-oplog-disabled-cold-transition' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'improvement/BB-460-Check-error-code-before-getting-metadata' into tmp/octopus/w/8.7/improvement/BB-460-Check-error-code-before-getting-metadata

bert-e committed:
Merge branches 'w/8.7/improvement/BB-460-Check-error-code-before-getting-metadata' and 'q/2462/8.6/improvement/BB-460-Check-error-code-before-getting-metadata' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/bugfix/BB-462/lifecycle' into tmp/octopus/w/8.7/bugfix/BB-462/lifecycle

bert-e committed:
Merge branch 'bugfix/BB-463-change-connector-partition-name-on-restart' into tmp/octopus/w/8.7/bugfix/BB-463-change-connector-partition-name-on-restart

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-463-change-connector-partition-name-on-restart' and 'q/2465/8.6/bugfix/BB-463-change-connector-partition-name-on-restart' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'bugfix/BB-441' into tmp/octopus/w/8.7/bugfix/BB-441

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-441' and 'q/2463/8.6/bugfix/BB-441' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'improvement/BB-417' into q/8.7

bert-e committed:
Merge branch 'bugfix/BB-440' into tmp/octopus/w/8.7/bugfix/BB-440

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-440' and 'q/2466/8.6/bugfix/BB-440' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'bugfix/BB-449' into tmp/octopus/w/8.7/bugfix/BB-449

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-449' and 'q/2467/8.6/bugfix/BB-449' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'bugfix/BB-455-clean-shutdown-backbeat-consumer' into tmp/octopus/w/8.7/bugfix/BB-455-clean-shutdown-backbeat-consumer

bert-e committed:
Merge branches 'w/8.7/bugfix/BB-455-clean-shutdown-backbeat-consumer' and 'q/2458/8.6/bugfix/BB-455-clean-shutdown-backbeat-consumer' into tmp/octopus/q/8.7

bert-e committed:
Merge branch 'w/8.6/feature/BB-452-delete-notification-versionid' into tmp/octopus/w/8.7/feature/BB-452-delete-notification-versionid

francoisferrand committed:
Fix jest lint warning

Issue: BB-295

francoisferrand committed:
Reenable bucket conductor test with bucketd

Issue: BB-295

francoisferrand committed:
Reenable disabled tess

Issue: BB-295
